tools/diff-kernel-config

```diff
@@ -0,0 +1,241 @@
+#
+
+mkdir -p "${output_dir}" || bail "Failed to create output directory '${output_dir}'"
+
+for state in before after; do
+
+    gitrev_var=gitrev_${state}
+    git checkout --quiet "${!gitrev_var}" || bail "Cannot check out '${!gitrev_var}'."
+
+    for arch in aarch64 x86_64; do
+
+        for kver in "${kernel_versions[@]}"; do
+
+            variants=()
+            case ${kver} in
+                5.10)
+                    variants+=( 'aws-k8s-1.23' )
+                    if [[ ${arch} = x86_64 ]]; then
+                        variants+=( 'metal-k8s-1.23' )
+                    fi
+                    ;;
+                5.15)
+                    variants+=( 'aws-dev' 'metal-dev' )
+                    ;;
+                *)
+                    bail "No known variants build kernel ${kver}."
+                    ;;
+            esac
+
+            for variant in "${variants[@]}"; do
+
+                debug_id="state=${state} arch=${arch} variant=${variant}"
+
+                #
+                # Run build
+                #
+
+                cargo make \
+                        -e BUILDSYS_ARCH="${arch}" \
+                        -e BUILDSYS_VARIANT="${variant}" \
+                        -e PACKAGE="kernel-${kver/./_}" \
+                        build-package \
+                    || bail "Build failed for ${debug_id}"
+
+                #
+                # Find kernel RPM
+                #
+
+                shopt -s nullglob
+                kernel_rpms=(
+                    ./build/rpms/bottlerocket-"${arch}"-*kernel-"${kver}"-"${kver}".*.rpm
+                )
+                shopt -u nullglob
+
+                case ${#kernel_rpms[@]} in
+                    0) bail "No kernel RPM found for ${debug_id}" ;;
+                    1) kernel_rpm=${kernel_rpms[0]} ;;
+                    *) bail "More than one kernel RPM found for ${debug_id}" ;;
+                esac
+
+
+                #
+                # Extract kernel config
+                #
+
+                config_path=./${output_dir}/config-${arch}-${kver}-${variant}-${state}
+                rpm2cpio "${kernel_rpm}" \
+                    | cpio --quiet --extract --to-stdout ./boot/config >"${config_path}"
+                [[ -s "${config_path}" ]] || bail "Failed to extract config for ${debug_id}"
+
+            done  # variant
+
+        done  # kver
+
+    done  # arch
+
+done  # state
+
+
+#
+# Post-process the collected pairs of "before" and "after" configs (generate diffs, a report, a summary)
+#
+
+# Get the helpful diffconfig script from the kernel source tree. We package it
+# in the kernel-archive RPM from where it can be extracted. Here we extract the
+# latest version of the script, but any kernel version and arch will do.
+latest_kver=$(printf '%s\n' "${kernel_versions[@]}" | sort -V | tail -n1)
+latest_archive_rpms=( ./build/rpms/bottlerocket-aarch64-kernel-"${latest_kver}"-archive-*.rpm )
+diffconfig=$(mktemp --suffix -bottlerocket-diffconfig)
```

Comment:
nit: would be good to add cleanup handler (e.g. `trap 'cleanup' EXIT`) to ensure the temp file doesn't leak

Reply:
Indeed. I originally planned on adding it for the happy path and littering /tmp for the unlikely failure path, but eventually even forgot about the former. To repent, I fixed both in the current revision. :smile: